### PR TITLE
feat: HUMSIG-37: ML backend experience updates

### PR DESCRIPTION
**Pull request by niklub**
- Support for HTTP Basic Authentication in ML Backend and Label Studio
- Capability to transfer additional JSON params from Label Studio to the ML Backend
- Enhanced configuration method for selecting predictions or ML Models in Label Studio for prelabelling
- Project settings featuring a list of predictions with options to remove them
- Standard screens for scenarios without connected models or uploaded predictions

## Old interface
image
image

## New interface
image
image

### CODE REVIEW

File: label_studio/data_manager/actions/basic.py
Comment (bmartel) [inline label_studio/data_manager/actions/basic.py]:
```suggestion
            'title': 'Retrieve Predictions',
```

File: label_studio/data_manager/actions/predictions_to_annotations.py
Comment (bmartel) [inline label_studio/data_manager/actions/predictions_to_annotations.py]:
```suggestion
            'title': 'Create Annotations From Predictions',
```

File: label_studio/tasks/api.py
Comment (jombooth) [inline label_studio/tasks/api.py]: why is this commented out? If it's no longer needed, let's delete it. But we should understand what it was doing before

File: label_studio/tasks/api.py
Comment (jombooth) [inline label_studio/tasks/api.py]: Why do we no longer check for `project.evaluate_predictions_automatically`?

File: label_studio/tasks/api.py
Comment (jombooth) [inline label_studio/tasks/api.py]: this looks like it can be deleted

File: .pre-commit-dev.yaml
Comment (jombooth) [inline .pre-commit-dev.yaml]: please revert this, it spams my terminal when I run `make fmt-all`

File: label_studio/data_manager/api.py
Comment (jombooth) [inline label_studio/data_manager/api.py]: What's the outcome of the discussion here? looks like another place where we used to consider `project.retrieve_predictions_automatically` but don't anymore. In the final version, we shouldn't have commented out code here.

File: label_studio/data_manager/api.py
Comment (jombooth) [inline label_studio/data_manager/api.py]: Similarly, we shouldn't merge commented out code here. What's the intent of this part of the change?

File: label_studio/ml/api.py
Comment (jombooth) [inline label_studio/ml/api.py]: We can delete this, or better yet, add a docstring :)

Comment (jombooth) [reply label_studio/ml/api.py]: (applies to all the `""" """` docstrings below also)

File: label_studio/tests/predictions/test_predictions.py
Comment (jombooth) [inline label_studio/tests/predictions/test_predictions.py]: please remove this print

File: label_studio/data_manager/functions.py
Comment (jombooth) [inline label_studio/data_manager/functions.py]: this is a small table so it likely doesn't matter, but it's generally better to use `fast_first` if you don't care about ordering

File: label_studio/data_manager/functions.py
Comment (jombooth) [inline label_studio/data_manager/functions.py]: I think this belongs in the docstring. What I'd recommend:

```
"""
Call the given ML backend to retrieve predictions with the task queryset as an input.
If backend is not specified, we'll assume the tasks' project only has one associated
ML backend, and use that backend.
"""
```

File: label_studio/ml/api.py
Comment (jombooth) [inline label_studio/ml/api.py]: Is this still a TODO? I can't really tell what's meant by the "this"

File: label_studio/tasks/models.py
Comment (jombooth) [inline label_studio/tasks/models.py]: The method shouldn't raise an exception for this. `project` should be a required argument, and there shouldn't be any nullchecking.

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: We generally discourage the use of Django signals because they aren't called under various conditions, eg, a queryset-level delete. why is this necessary? A docstring would really help.

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: please use update_fields to avoid the potential for clobbering other updates to `project`. like `project.save(update_fields=['model_version'])`

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: this should also use update_fields as recommended in https://github.com/HumanSignal/label-studio/pull/5570/files#r1522245983

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: It's really better to avoid custom `save()` methods if at all possible. Is there a different way we could do this? For example, could we just make the project's model_version optional and defer to the title of the model instance if it is unset?

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: nit: please add a docstring or remove this

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: We save almost nothing by using two-letter codes in this charfield, why not make the CharField len 255 and make the enum values the same as the enum name? This means nobody ever has to figure out that "NA" is the same as NONE by reading backend code, for example.

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py:79]: On recent versions of postgres, adding a column with a simple default is an efficient operation. I would have preferred specifying either `null=True` or `default=''` but not both; however I recognize the ship may have already sailed on this since it was deployed on a production database.

File: label_studio/ml/serializers.py
Comment (jombooth) [inline label_studio/ml/serializers.py:101]: We need to make sure that these fields are writable but not readable in the API.

File: label_studio/projects/models.py
Comment (jombooth) [inline label_studio/projects/models.py]: need to remove this too

File: label_studio/projects/models.py
Comment (jombooth) [inline label_studio/projects/models.py]: please use `fast_first` instead of `first()`

File: label_studio/projects/models.py
Comment (jombooth) [inline label_studio/projects/models.py]: We should remove this comment

File: label_studio/projects/models.py
Comment (jombooth) [inline label_studio/projects/models.py]: we should be using update_fields here too

File: label_studio/projects/models.py
Comment (jombooth) [inline label_studio/projects/models.py]: optional: I believe the delete() method returns the number of rows that it deleted, so we can ditch the count() query

File: label_studio/projects/serializers.py
Comment (jombooth) [inline label_studio/projects/serializers.py]: This comment seems to be outdated, or is `'show_collab_predictions'` wrong below?

File: label_studio/tests/ml/test_models.py
Comment (jombooth) [inline label_studio/tests/ml/test_models.py]: please remove this and other empty files from the diff

File: label_studio/projects/api.py
Comment (jombooth) [inline label_studio/projects/api.py:646]: I think this should be addressed before merging

File: label_studio/ml/serializers.py
Comment (jombooth) [reply label_studio/ml/serializers.py:101]: ok, I do see this is write_only: https://github.com/HumanSignal/label-studio/pull/5570/files#diff-a67b82b0de34ce236f0f9a212e530d2a865c73f603e0e5e6ca485d5c0ea45d10R13

would be great to capture this in a test though.

File: label_studio/ml/api_connector.py
Comment (jombooth) [inline label_studio/ml/api_connector.py]: Here and everywhere in the PR: we shouldn't refer to strings from an enum explicitly. Instead, let's compare `auth_method` with `MLBackendAuth.BASIC_AUTH`

Ideally we would use an enum on the frontend too for the same reason.

File: label_studio/projects/models.py
Comment (jombooth) [inline label_studio/projects/models.py]: I don't see `*args` or `**kwargs` being used anywhere

File: label_studio/tasks/api.py
Comment (jombooth) [reply label_studio/tasks/api.py]: I have since learned that it was deprecated as part of this work - we should add this fact to the PR description

File: label_studio/ml/models.py
Comment (niklub) [inline label_studio/ml/models.py]: let's fix this

File: label_studio/ml/models.py
Comment (niklub) [inline label_studio/ml/models.py]: maybe rollack it again

File: label_studio/ml/serializers.py
Comment (niklub) [inline label_studio/ml/serializers.py:15]: adding test to check GET payload

File: label_studio/data_manager/functions.py
Comment (niklub) [reply label_studio/data_manager/functions.py]: since it is only one backend connected all the time, I wouldn't introduce this change rn to avoid big diff

File: label_studio/projects/serializers.py
Comment (jombooth) [inline label_studio/projects/serializers.py]: I think you could do (pseudocode)

```
MlBackend.objects.filter(project=p, title=value).union(Prediction.objects.filter(project=p, model_version=value)).exists()
```

this would be one query instead of four, assuming p.ml_backends and p.predictions aren't prefetched.

File: label_studio/tests/next_task.tavern.yml
Comment (jombooth) [inline label_studio/tests/next_task.tavern.yml:119]: what's behind the changes to this test?

File: label_studio/ml/api.py
Comment (jombooth) [inline label_studio/ml/api.py]: There are three test failures coming from this code, because `ml_api_result` at the top is a dictionary, not an `MLApiResult`, and therefore doesn't have any members like `is_error`. The tests that need to be fixed are:

```
tests/test_predictions.py::test_interactive_annotating
tests/test_predictions.py::test_interactive_annotating_failing
tests/test_predictions.py::test_interactive_annotating_with_drafts
```

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py:138]: I think to be comfortable merging this, we really need a testcase that captures the behavior of this function (ideally something relatively high-level demonstrating the sequence of realistic API calls that causes project.model_version to be renamed, not just a unit test)

File: label_studio/ml/models.py
Comment (jombooth) [inline label_studio/ml/models.py]: as with MLBackend#save above, we really need a testcase that, using realistic API calls, captures the impact that this signal handler has.

**Comment by jombooth** [comment]
Requesting screenshots of all new in-app functionality in the description, two testcases for model_version update behavior on save and delete, and the GET api test for basic_auth_pass mentioned earlier in the review - with these changes, I'm prepared to approve for the backend

File: label_studio/projects/serializers.py
Comment (jombooth) [inline label_studio/projects/serializers.py]: please remove commented out code

**Comment by jombooth** [comment]
Thanks for adding old vs new screenshots, super helpful :+1:

File: label_studio/ml/api_connector.py
Comment (jombooth) [inline label_studio/ml/api_connector.py]: What does this mean? Please clarify.

**Review by jombooth — APPROVED** [review]
Approving for backend changes, please address remaining comments about TODOs etc

**Review by bmartel — APPROVED** [review]
Approving for frontend. Thanks @niklub @jombooth 🎉

File: label_studio/projects/api.py
Comment (niklub) [reply label_studio/projects/api.py:646]: I think we don't have consistent naming convention regarding request args

**Review by jombooth — APPROVED** [review]
LGTM, also fixed ruff

**Pull request merged**
